File: library/src/main/java/com/schibsted/spain/barista/BaristaAssertions.java
package com.schibsted.spain.barista;

import android.support.annotation.IdRes;
import android.support.annotation.StringRes;
import android.support.test.espresso.NoActivityResumedException;
import android.util.Log;
import com.schibsted.spain.barista.androidresource.ResourceTypeChecker;
import com.schibsted.spain.barista.exception.BaristaArgumentTypeException;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.Espresso.pressBack;
import static android.support.test.espresso.assertion.ViewAssertions.doesNotExist;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.contrib.DrawerMatchers.isClosed;
import static android.support.test.espresso.contrib.DrawerMatchers.isOpen;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.isEnabled;
import static android.support.test.espresso.matcher.ViewMatchers.withHint;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static com.schibsted.spain.barista.custom.HelperMatchers.firstViewOf;
import static org.hamcrest.Matchers.allOf;
import static org.hamcrest.core.IsNot.not;
import static org.junit.Assert.fail;

public class BaristaAssertions {

  private static final ResourceTypeChecker RESOURCE_TYPE_CHECKER = new ResourceTypeChecker();

  public static void assertDisplayed(int id) {
    if (isIdResource(id)) {
      onView(withId(id)).check(matches(isDisplayed()));
    } else if (isStringResource(id)) {
      onView(withText(id)).check(matches(isDisplayed()));
    } else {
      throw new BaristaArgumentTypeException();
    }
  }

  public static void assertDisplayed(String text) {
    onView(firstViewOf(allOf(withText(text), isDisplayed()))).check(matches(isDisplayed()));
  }

  public static void assertNotExist(int id) {
    if (isIdResource(id)) {
      onView(withId(id)).check(doesNotExist());
    } else if (isStringResource(id)) {
      onView(withText(id)).check(doesNotExist());
    } else {
      throw new BaristaArgumentTypeException();
    }
  }

  public static void assertNotExist(String text) {
    onView(withText(text)).check(doesNotExist());
  }

  public static void assertNotDisplayed(int id) {
    if (isIdResource(id)) {
      onView(withId(id)).check(matches(not(isDisplayed())));
    } else if (isStringResource(id)) {
      onView(withText(id)).check(matches(not(isDisplayed())));
    } else {
      throw new BaristaArgumentTypeException();
    }
  }

  public static void assertNotDisplayed(String text) {
    onView(withText(text)).check(matches(not(isDisplayed())));
  }

  public static void assertEnabled(int id) {
    if (isIdResource(id)) {
      onView(withId(id)).check(matches(isEnabled()));
    } else if (isStringResource(id)) {
      onView(withText(id)).check(matches(isEnabled()));
    } else {
      throw new BaristaArgumentTypeException();
    }
  }

  public static void assertEnabled(String text) {
    onView(withText(text)).check(matches(isEnabled()));
  }

  public static void assertDisabled(int id) {
    if (isIdResource(id)) {
      onView(withId(id)).check(matches(not(isEnabled())));
    } else if (isStringResource(id)) {
      onView(withText(id)).check(matches(not(isEnabled())));
    } else {
      throw new BaristaArgumentTypeException();
    }
  }

  public static void assertDisabled(String text) {
    onView(withText(text)).check(matches(not(isEnabled())));
  }

  public static void assertThatBackButtonClosesTheApp() {
    try {
      pressBack(); // Will launch an Exception if it closes the app
      fail(); // One of our Activities is appearing on the screen :(
    } catch (NoActivityResumedException expectedException) {
      Log.d("Barista",
          "As the Activity is the first one of the stack, we expected this error. Yes, the back button closes the app!");
    }
  }

  public static void assertDrawerIsOpen(@IdRes int id) {
    onView(withId(id)).check(matches(isOpen()));
  }

  public static void assertDrawerIsClosed(@IdRes int id) {
    onView(withId(id)).check(matches(isClosed()));
  }

  public static void assertHint(@IdRes int id, @StringRes int text) {
    onView(withId(id)).check(matches(withHint(text)));
  }

  public static void assertHint(@IdRes int id, String text) {
    onView(withId(id)).check(matches(withHint(text)));
  }

  private static boolean isIdResource(int id) {
    return RESOURCE_TYPE_CHECKER.isIdResource(id);
  }

  private static boolean isStringResource(int id) {
    return RESOURCE_TYPE_CHECKER.isStringResource(id);
  }
}

File: library/src/main/java/com/schibsted/spain/barista/BaristaAssertions.kt
package com.schibsted.spain.barista

import android.support.annotation.DrawableRes
import android.support.annotation.IdRes
import android.support.annotation.StringRes
import android.support.test.espresso.AmbiguousViewMatcherException
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.NoActivityResumedException
import android.support.test.espresso.NoMatchingViewException
import android.support.test.espresso.action.ViewActions
import android.support.test.espresso.assertion.ViewAssertions.doesNotExist
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.contrib.DrawerMatchers.isClosed
import android.support.test.espresso.contrib.DrawerMatchers.isOpen
import android.support.test.espresso.matcher.ViewMatchers.*
import android.support.v4.view.GravityCompat
import android.view.View
import com.schibsted.spain.barista.custom.DisplayedMatchers.displayedWithId
import com.schibsted.spain.barista.custom.DrawableMatchers.Companion.withDrawable
import com.schibsted.spain.barista.custom.HelperMatchers.firstViewOf
import com.schibsted.spain.barista.custom.RecyclerViewItemCountAssertion
import com.schibsted.spain.barista.internal.failurehandler.RethrowingFailureHandler
import com.schibsted.spain.barista.internal.failurehandler.SpyFailureHandler
import com.schibsted.spain.barista.internal.failurehandler.description
import com.schibsted.spain.barista.internal.resourceMatcher
import org.hamcrest.Matcher
import org.hamcrest.Matchers.allOf
import org.hamcrest.core.IsNot.not
import org.junit.Assert.fail

object BaristaAssertions {

    @JvmStatic
    fun assertDisplayed(resId: Int) {
        assertDisplayed(resId.resourceMatcher())
    }

    @JvmStatic
    fun assertDisplayed(text: String) {
        assertDisplayed(withText(text))
    }

    private fun assertDisplayed(matcher: Matcher<View>) {
        val spyFailureHandler = SpyFailureHandler()
        try {
            onView(matcher)
                    .withFailureHandler(spyFailureHandler)
                    .check(matches(isDisplayed()))
        } catch (multipleViews: AmbiguousViewMatcherException) {
            try {
                onView(firstViewOf(allOf(matcher, isDisplayed())))
                        .withFailureHandler(spyFailureHandler)
                        .check(matches(isDisplayed()))
            } catch (notDisplayedError: NoMatchingViewException) {
                spyFailureHandler.resendFirstError("View ${matcher.description()} wasn't displayed on the screen")
            }
        }
    }

    @JvmStatic
    fun assertNotExist(resId: Int) {
        onView(resId.resourceMatcher()).check(doesNotExist())
    }

    @JvmStatic
    fun assertNotExist(text: String) {
        onView(withText(text)).check(doesNotExist())
    }

    @JvmStatic
    fun assertNotDisplayed(resId: Int) {
        onView(resId.resourceMatcher()).check(matches(not(isDisplayed())))
    }

    @JvmStatic
    fun assertNotDisplayed(text: String) {
        onView(withText(text)).check(matches(not(isDisplayed())))
    }

    @JvmStatic
    fun assertEnabled(resId: Int) {
        onView(resId.resourceMatcher()).check(matches(isEnabled()))
    }

    @JvmStatic
    fun assertEnabled(text: String) {
        onView(withText(text)).check(matches(isEnabled()))
    }

    @JvmStatic
    fun assertDisabled(resId: Int) {
        onView(resId.resourceMatcher()).check(matches(not(isEnabled())))
    }

    @JvmStatic
    fun assertDisabled(text: String) {
        onView(withText(text)).check(matches(not(isEnabled())))
    }

    @JvmStatic
    fun assertChecked(resId: Int) {
        onView(resId.resourceMatcher()).check(matches(isChecked()))
    }

    @JvmStatic
    fun assertChecked(text: String) {
        onView(withText(text)).check(matches(isChecked()))
    }

    @JvmStatic
    fun assertUnchecked(resId: Int) {
        onView(resId.resourceMatcher()).check(matches(isNotChecked()))
    }

    @JvmStatic
    fun assertUnchecked(text: String) {
        onView(withText(text)).check(matches(isNotChecked()))
    }

    @JvmStatic
    fun assertThatBackButtonClosesTheApp() {
        // Will launch an Exception if it closes the app
        try {
            onView(isRoot())
                    .withFailureHandler(RethrowingFailureHandler())
                    .perform(ViewActions.pressBack())
            // One of our Activities is appearing on the screen :(
            fail("The back button didn't close the app")
        } catch (noActivityException: NoActivityResumedException) {
            //Yey!
        }
    }

    @JvmStatic
    fun assertRecyclerViewItemCount(@IdRes recyclerViewId: Int, expectedItemCount: Int) {
        onView(displayedWithId(recyclerViewId)).check(
                RecyclerViewItemCountAssertion(expectedItemCount))
    }

    @JvmStatic
    fun assertDrawerIsOpen(@IdRes id: Int) {
        assertStartDrawerOpen(id)
    }

    @JvmStatic
    fun assertDrawerIsClosed(@IdRes id: Int) {
        assertStartDrawerClosed(id)
    }

    @JvmStatic
    fun assertStartDrawerOpen(@IdRes id: Int) {
        onView(withId(id)).check(matches(isOpen(GravityCompat.START)))
    }

    @JvmStatic
    fun assertStartDrawerClosed(@IdRes id: Int) {
        onView(withId(id)).check(matches(isClosed(GravityCompat.START)))
    }

    @JvmStatic
    fun assertEndDrawerOpen(@IdRes id: Int) {
        onView(withId(id)).check(matches(isOpen(GravityCompat.END)))
    }

    @JvmStatic
    fun assertEndDrawerClosed(@IdRes id: Int) {
        onView(withId(id)).check(matches(isClosed(GravityCompat.END)))
    }

    @JvmStatic
    fun assertHint(@IdRes id: Int, @StringRes text: Int) {
        onView(withId(id)).check(matches(withHint(text)))
    }

    @JvmStatic
    fun assertHint(@IdRes id: Int, text: String) {
        onView(withId(id)).check(matches(withHint(text)))
    }

    @JvmStatic
    fun assertDrawable(@IdRes id: Int, @DrawableRes drawable: Int) {
        onView(withId(id)).check(matches(withDrawable(drawable)))
    }
}
